<launch>

  <arg name="robot_name" default="dragonfly11"/>
  <arg name="mass" default="0.250"/>
	<arg name="vicon" default="0"/>
	<arg name="vicon_fps" default="100"/>
	<arg name="start_nodelet_manager" default="1"/>
	<arg name="nodelet_manager_name" default="nodelet_manager_control"/>
  <arg name="sim" default="0"/>
  <arg name="odotopic" value="/$(arg robot_name)/quadrotor_ukf/control_odom" unless="$(arg vicon)"/>
  <arg name="odotopic" value="/vicon/$(arg robot_name)/odom" if="$(arg vicon)"/>

    <group ns="$(arg robot_name)">

      <param name="mass" value="$(arg mass)"/>
	    
	    <!-- Master Sync Stuff -->
	    <node pkg="master_discovery_fkie" type="master_discovery" name="master_discovery">
	      <param name="mcast_group" value="224.0.0.1" />
	    </node>

	    <node pkg="master_sync_fkie" type="master_sync" name="multimaster_client_sync">
	      <rosparam file="$(find control_cctv)/config/multimaster_params.yaml" />
	    </node>

	    <!-- Nodelet manager-->
	    <node pkg="nodelet"
	      type="nodelet"
        name="standalone_nodelet"
	      args="manager"
	      output="screen"
        if="$(arg start_nodelet_manager)">
        <rosparam file="$(find mav_manager)/config/mav_manager_params.yaml" />
      </node>

	    <!-- MAV Manager -->
	    <node pkg="mav_manager"
	      type="mav_services"
	      name="mav_services"
	      clear_params="true"
	      output="screen">
	      <rosparam file="$(find mav_manager)/config/mav_manager_params.yaml" />
	      <remap from="~odom" to="$(arg odotopic)"/>
	      <remap from="odom" to="$(arg odotopic)"/>
	      <remap from="~des_odom" to="des_odom"/>
	    </node>

	    <!-- Trackers Manager -->
	    <node pkg="nodelet"
	      type="nodelet"
	      name="trackers_manager"
	      args="load trackers_manager/TrackersManager standalone_nodelet"
	      output="screen">
	      <rosparam file="$(find trackers_manager)/config/trackers.yaml"/>
	      <rosparam file="$(find trackers_manager)/config/tracker_params.yaml"/>
	      <rosparam file="$(find so3_control)/config/gains.yaml"/>
        <remap from="~odom" to="$(arg odotopic)"/>
	      <remap from="~cmd" to="position_cmd"/>
	    </node>

	    <!--CCTV Control nodelet-->
	    <node pkg="nodelet"
	      type="nodelet"
	      args="load control_cctv/NodeletControlCCTV  standalone_nodelet"
	      name="control_cctv"
	      required="true"
	      output="screen">
	      <rosparam file="$(find control_cctv)/config/cable_vision_params.yaml"/>
	      <param name="mass" value="$(arg mass)"/>
        <remap from="~quad_odom" to="$(arg odotopic)"/>
	      <remap from="~payload_odom" to="pl_pose"/>
	      <remap from="~position_cmd" to="position_cmd"/>
	      <remap from="~so3_cmd" to="so3_cmd"/>
	      <remap from="~motors" to="motors"/>
      </node>

      <node pkg="nodelet"
        type="nodelet"
        name="quad_encode_msg"
        args="load quad_encode_msg/QuadEncodeMsg standalone_nodelet"
        output="screen"
        unless="$(arg sim)">
        <remap from="~so3_cmd" to="so3_cmd"/>
        <remap from="~serial_msg" to="to_robot"/>
      </node>

      <node pkg="nodelet"
        type="nodelet"
          args="standalone qualcomm_interface/SO3CmdToQualcomm"
          name="so3cmd_to_qualcomm_nodelet"
          required="true"
          clear_params="true"
          output="screen"
          unless="$(arg sim)">
          <remap from="~odom" to="$(arg odotopic)"/>
          <remap from="~so3_cmd" to="so3_cmd"/>
      </node>

	    <group unless="$(arg vicon)">

		    <node pkg="vio_qc"
	    	    type="MVSampleVISLAMEagle"
			    name="MVSampleVISLAMEagle"
		        output="screen" args = "-a">
		        <remap from="~vio_pose" to="vio_pose"/>
		        <remap from="~imu" to="imu"/>
		    </node>

	    	<node pkg="quadrotor_ukf"
	  		    type="quadrotor_ukf"
		        name="quadrotor_ukf"
		        output="screen">
		        <remap from="~odom_slam" to="vio_pose"/>
		        <remap from="~imu" to="imu"/>
	      	</node>

        <node pkg="topic_tools"
            type="throttle"
            name="odom_throttle"
            args="messages $(arg odotopic) 10 odom_throttled">
        </node>
	    </group>
	</group>
</launch>
